src/shutdown/shutdown.c

```diff
@@ -531,10 +531,6 @@ int main(int argc, char *argv[]) {
                           need_md_detach ? " MD devices," : "");
         }
 
-        /* We're done with the watchdog. */
-        watchdog_close(true);
-        watchdog_free_device();
```

Comment:
I think we should keep the watchdog_free_device() in place, so that we release all resources. I mean, the very likely case is that we'll execve() very soon anyway, and that will close the fd anyway.

The kernel watchog API is explicitly document to live the watchdog "hot" if you just close the device. If you want to disarm it you have to write "V" to it, iirc first, and we do that whenever we want to close the device without leaving it armed. hence: please drop the watchdog_close(), i.e. revert https://github.com/systemd/systemd/pull/22079 please, but do not also drop watchdog_free_device()...

In general it's nice if our code remains "valgrind" clean, even if IRL this doesn#t matter. But it makes our codebase easier to manage and validate and test. 

As far as I can see the only benefit of removing that line is that in case no "exitrd" is used (i.e not on the big distros) that one kernel log line is suppressed, but that sounds like a pretty weak improvement to me. I'd rather see it removed from the kernel, given we knowingly leave the watchdog armed here (well, at least until #22079 was merged, but that should be reverted, as said).

We should also add a comment here, explaining things, so that this doesn't happen again.

Reply:
Changed it that way and added a comment.